### PR TITLE
[ALLUXIO-3301] S3A Streaming upload

### PR DESCRIPTION
**Pull request by LuQQiu**
https://alluxio.atlassian.net/browse/ALLUXIO-3301

### CODE REVIEW

**Comment by LuQQiu** [comment]
jenkins, test this please

**Comment by LuQQiu** [comment]
@madanadit @calvinjia PTAL, thanks!
Fulfill the main logic, now working on testing it and adding unit/integration tests.

File: core/client/fs/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java [hunk 1]
```diff
@@ -110,6 +113,9 @@
   /** This condition is met if there is nothing in the netty buffer. */
   private final Condition mBufferEmptyOrFailed = mLock.newCondition();
 
+  /** Flush should not be called before write. */
+  private boolean mWrote = false;
```
Comment (calvinjia) [inline core/client/fs/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java]: Could we allow the RPC to go through and no-op on the server side? This would simplify the client state.

File: core/client/fs/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java [hunk 2]
```diff
@@ -481,13 +489,13 @@ public void operationComplete(ChannelFuture future) {
   }
 
   /**
-   * The netty channel future listener that is called when a EOF or CANCEL is complete.
+   * The netty channel future listener that is called when a EOF or CANCEL or FLUSH is complete.
    */
-  private final class EofOrCancelListener implements ChannelFutureListener {
+  private final class EofOrCancelorFlushListener implements ChannelFutureListener {
```
Comment (calvinjia) [inline core/client/fs/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java]: This listener cannot be reused because `Cancel` and `EOF` are terminal RPCs whereas `flush` is not

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioFuseFileSystem.java [hunk 3]
```diff
@@ -762,6 +771,30 @@ private int rmInternal(String path, boolean mustBeFile) {
...
+        }
+      }, WaitForOptions.defaults().setTimeoutMs(MAX_OPEN_WAITTIME));
+      return true;
+    } catch (InterruptedException ie) {
+      Thread.currentThread().interrupt();
+      return false;
+    } catch (TimeoutException te) {
+      return false;
+    }
+  }
```
Comment (calvinjia) [inline integration/fuse/src/main/java/alluxio/fuse/AlluxioFuseFileSystem.java]: If it is not too much trouble, could you put the FUSE changes in a separate PR? These seem like independent changes.

File: core/common/src/main/java/alluxio/heartbeat/HeartbeatContext.java [hunk 4]
```diff
@@ -39,6 +39,8 @@
   public static final String MASTER_LOST_MASTER_DETECTION = "Master Lost Master Detection";
   public static final String MASTER_LOST_WORKER_DETECTION = "Master Lost Worker Detection";
   public static final String MASTER_METRICS_SYNC = "Master Metrics Sync";
+  public static final String MASTER_S3A_INTERMEDIATE_UPLOAD_CLEAN
+      = "Master S3A Intermediate Upload Clean";
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/heartbeat/HeartbeatContext.java]: I would suggest to make this agnostic to the UFS, ie. `Master UFS Cleanup` and call a `cleanup` method for each of the registered UnderFileSystems which for S3 does the existing logic. @madanadit what do you think about adding `cleanup` to the UFS interface?

**Comment by LuQQiu** [comment]
jenkins, test this please

File: core/server/master/src/main/java/alluxio/master/file/S3AUploadCleaner.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * S3A intermediate multipart upload periodic clean.
+ */
+@NotThreadSafe
+final class S3AUploadCleaner implements HeartbeatExecutor {
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/file/S3AUploadCleaner.java]: Would it be cleaner (no pun intended) to put this logic in the S3 under filesystem?

Comment (LuQQiu) [reply core/server/master/src/main/java/alluxio/master/file/S3AUploadCleaner.java]: In the previous meeting, we thought it may be too many cleanup if we put it in the S3 under filesystem. But it would be definitely cleaner to put this logic in the S3AUnderFileSystem, we do not need to create additional transfer managers and could reuse nearly all of the codes.

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioFuseFileSystem.java (hunk 3 above)
Comment (LuQQiu) [reply integration/fuse/src/main/java/alluxio/fuse/AlluxioFuseFileSystem.java]: Sure, I would put Fuse changes in another PR, thanks!

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java [hunk 6]
```diff
@@ -222,16 +233,21 @@ public static S3AUnderFileSystem createInstance(AlluxioURI uri,
    * @param uri the {@link AlluxioURI} for this UFS
    * @param amazonS3Client AWS-SDK S3 client
    * @param bucketName bucket name of user's configured Alluxio bucket
+   * @param executor the executor for executing upload tasks
    * @param transferManager Transfer Manager for efficient I/O to S3
    * @param conf configuration for this S3A ufs
+   * @param streamingUploadEnabled whether streaming uplaod is enabled
```
Comment (uronce-cc) [inline underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java]: NIT: typo, uplaod -> upload

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * [Experimental] A stream for writing a file into S3 using streaming upload.
+ * The data will be persisted to a temporary directory on the local disk
+ * and uploaded when the buffered data reaches the partition size. The data
+ * transfer is done using an S3 low-level multipart upload API.
+ */
+@NotThreadSafe
+public class S3ALowLevelOutputStream extends OutputStream {
```
Comment (uronce-cc) [inline underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java]: renamed to S3AStreamingOutputStream?

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
+  private static final Logger LOG = LoggerFactory.getLogger(S3ALowLevelOutputStream.class);
+
+  private static final boolean SSE_ENABLED =
+      Configuration.getBoolean(PropertyKey.UNDERFS_S3A_SERVER_SIDE_ENCRYPTION_ENABLED);
+
+  /**
+   * Only parts bigger than 5MB could be uploaded through multipart upload,
+   * except the last part.
+   */
+  private static final long UPLOAD_THRESHOLD = 5 * Constants.MB;
```
Comment (uronce-cc) [inline underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java]: make this a configurable property in case users want to increase the threshold?

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java [hunk 9]
```diff
@@ -0,0 +1,406 @@
...
+      ExecutorService executor) {
+    Preconditions.checkArgument(bucketName != null && !bucketName.isEmpty(), "Bucket name must "
+        + "not be null or empty.");
+    mBucketName = bucketName;
+    mClient = s3Client;
+    mExecutor = MoreExecutors.listeningDecorator(executor);
+    try {
+      mHash = MessageDigest.getInstance("MD5");
+    } catch (NoSuchAlgorithmException e) {
+      LOG.warn("Algorithm not available for MD5 hash.", e);
```
Comment (uronce-cc) [inline underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java]: this should never happen, can throw RuntimeException

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java [hunk 10]
```diff
@@ -0,0 +1,406 @@
...
+      mHash = null;
+    }
+    mKey = key;
+    // Partition size should be at least 5 MB, since S3 low-level multipart upload does not
+    // accept intermediate part smaller than 5 MB.
+    long partSize = Configuration.getBytes(PropertyKey.UNDERFS_S3A_STREAMING_UPLOAD_PARTITION_SIZE);
+    mPartitionSize = partSize < 5 * Constants.MB ? 5 * Constants.MB : partSize;
```
Comment (uronce-cc) [inline underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java]: min(UPLOAD_THRESHOLD, partSize)

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java [hunk 11]
```diff
@@ -0,0 +1,406 @@
...
+   */
+  private void initMultiPartUpload() throws IOException {
+    // Generate the object metadata by setting server side encryption, md5 checksum,
+    // and encoding as octet stream since no assumptions are made about the file type
+    ObjectMetadata meta = new ObjectMetadata();
+    if (SSE_ENABLED) {
+      meta.setSSEAlgorithm(ObjectMetadata.AES_256_SERVER_SIDE_ENCRYPTION);
+    }
+    if (mHash != null) {
+      meta.setContentMD5(new String(Base64.encode(mHash.digest())));
```
Comment (uronce-cc) [inline underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java]: use `encodeAsString` in case the default system default charset decode the encoded bytes incorrectly in the constructor of `String`.

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java [hunk 12]
```diff
@@ -0,0 +1,406 @@
...
+        .withPartNumber(partNumber)
+        .withFile(newFileToUpload)
+        .withPartSize(newFileToUpload.length());
+    execUpload(uploadRequest, newFileToUpload, partNumber);
+  }
+
+  /**
+   * Executes the upload part request.
+   */
+  private void execUpload(UploadPartRequest request, File file, int partNumber) {
```
Comment (uronce-cc) [inline underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java]: `file` and `partNumber` can be retrieved from `request`

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java [hunk 13]
```diff
@@ -0,0 +1,406 @@
...
+    }
+  }
+
+  @Override
+  public void flush() throws IOException {
+    // We try to minimize the time use to close()
+    // because Fuse release() method which calls close() is async.
+    // In flush(), we upload the current writing file if it is bigger than 5 MB,
+    // and wait for all current upload to complete.
+    mLocalOutputStream.flush();
```
Comment (uronce-cc) [inline underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java]: If `flush` is called immediately after `write`, then `mLocalOutputStream` is `null`.

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java [hunk 14]
```diff
@@ -0,0 +1,406 @@
...
+        mLocalOutputStream.close();
+        int partNumber = mPartNumber.getAndIncrement();
+        final UploadPartRequest uploadRequest = new UploadPartRequest()
+            .withBucketName(mBucketName)
+            .withKey(mKey)
+            .withUploadId(mUploadId)
+            .withPartNumber(partNumber)
+            .withFile(mFile)
+            .withPartSize(mFile.length());
+        uploadRequest.setLastPart(true);
```
Comment (uronce-cc) [inline underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java]: if each time when write is called, length is exactly mPartitionSize, then when `close` is called, `mFile` is `null`, so `setLastPart(true)` will never be called, does `setLastPart` must be called for the last part?

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java [hunk 15]
```diff
@@ -0,0 +1,406 @@
...
+  /**
+   * Waits for the submitted upload tasks to finish.
+   */
+  private void waitForAllPartsUpload() throws IOException {
+    try {
+      mTags.addAll(Futures.allAsList(mFutureTags).get());
+    } catch (ExecutionException e) {
+      // No recover ways so that we need to cancel all the upload tasks
+      // and abort the multipart upload
+      for (ListenableFuture<PartETag> future : mFutureTags) {
```
Comment (uronce-cc) [inline underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java]: cancel the future returned by `addAll` will cancel all futures in mFutureTags.

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java [hunk 16]
```diff
@@ -0,0 +1,406 @@
...
+        future.cancel(true);
+      }
+      abortMultiPartUpload();
+      throw new IOException("Part upload failed in multipart upload with "
+          + "id '" + mUploadId + "' to " + mKey, e);
+    } catch (InterruptedException e) {
```
Comment (uronce-cc) [inline underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java]: if any one of the threads in the executor is interrupted, `allAsList(...).get()` will throw this exception, but other threads in the executor may still be running, so futures need to be cancelled here too.

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java [hunk 17]
```diff
@@ -222,16 +233,21 @@ public static S3AUnderFileSystem createInstance(AlluxioURI uri,
...
    */
   protected S3AUnderFileSystem(AlluxioURI uri, AmazonS3Client amazonS3Client, String bucketName,
-      TransferManager transferManager, UnderFileSystemConfiguration conf) {
+      ExecutorService executor, TransferManager transferManager, UnderFileSystemConfiguration conf,
+      boolean streamingUploadEnabled) {
     super(uri, conf);
     mClient = amazonS3Client;
     mBucketName = bucketName;
+    mExecutor = executor;
```
[the remark this answers is not in the document]
Comment (LuQQiu) [reply underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java]: All the underFileSystem resources will be cleaned at the end when we 'cleanupRequest' or 'cancelRequest' in the 'UfsWriteHandler'.

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java [hunk 18]
```diff
@@ -0,0 +1,406 @@
...
+    InitiateMultipartUploadRequest initRequest =
+        new InitiateMultipartUploadRequest(mBucketName, mKey).withObjectMetadata(meta);
+    do {
+      try {
+        mUploadId = mClient.initiateMultipartUpload(initRequest).getUploadId();
+        return;
+      } catch (AmazonClientException e) {
+        lastException = e;
+      }
+    } while (mRetryPolicy.attempt());
```
[the remark this answers is not in the document]
Comment (LuQQiu) [reply underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java]: I want the whole upload request can retry for 5 times instead of each operation could retry for 5 times. If the whole request occurs errors more than 5 times, we better to cancel it and clean up.

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java (hunk 9 above)
Comment (LuQQiu) [reply underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java]: Agree, this should never happen, but throw an exception in the constructor should be avoided.
Putting mHash initialization in other methods looks wired, I suggest remaining it here.

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java (hunk 7 above)
Comment (LuQQiu) [reply underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java]: We first come up with the `S3AStreamingOutputStream` name and thinks it's a little strange. Haven't come up with better names....

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java (hunk 16 above)
Comment (LuQQiu) [reply underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java]: Make sense, if interrupted, we should also cancel all the uploads and abort the whole multipart upload. Thanks!
```
    } catch (InterruptedException e) {
      LOG.warn("Interrupted object upload.", e);
      Futures.allAsList(mFutureTags).cancel(true);
      abortMultiPartUpload();
      Thread.currentThread().interrupt();
    }
```

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java (hunk 15 above)
Comment (LuQQiu) [reply underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java]: Thanks! Great and convenient methods.

Changed to
```
Futures.allAsList(mFutureTags).cancel(true);
```

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java (hunk 13 above)
Comment (LuQQiu) [reply underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java]: The mLocalOutputStream do have the chance to be null although it may be low.
Add a null check to make it more robust.
Changed to
```
    if (mLocalOutputStream != null) {
      mLocalOutputStream.flush();
    }
```
Thanks

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java (hunk 12 above)
Comment (LuQQiu) [reply underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java]: Removed those parameters, thanks

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java (hunk 11 above)
Comment (LuQQiu) [reply underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java]: Thanks! Changed.

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java (hunk 10 above)
Comment (LuQQiu) [reply underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java]: Changed, thanks!

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java (hunk 6 above)
Comment (LuQQiu) [reply underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java]: Fixed, thanks

File: core/client/fs/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java (hunk 1 above)
Comment (LuQQiu) [reply core/client/fs/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java]: Now the client side will send the flush command to worker side.
Worker side will not actually flush() if no write() occurred.

File: core/client/fs/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java (hunk 2 above)
Comment (LuQQiu) [reply core/client/fs/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java]: Thanks for the kind remind. Add a new `FlushListener` for flush only.

File: core/common/src/main/java/alluxio/heartbeat/HeartbeatContext.java (hunk 4 above)
Comment (LuQQiu) [reply core/common/src/main/java/alluxio/heartbeat/HeartbeatContext.java]: Create a `UnderFileSystem.cleanup()` method. And setup periodically cleanup all non-readonly ufs in `DefaultFileSystemMaster`
Thanks

File: core/server/master/src/main/java/alluxio/master/file/S3AUploadCleaner.java (hunk 5 above)
Comment (LuQQiu) [reply core/server/master/src/main/java/alluxio/master/file/S3AUploadCleaner.java]: Put the cleanup logic in the `S3AUnderFileSystem.cleanup`. Leader master will clean up all UFS periodically if `UNDERFS_CLEANUP_ENABLED`. This makes the code much cleaner! Thanks

**Comment by LuQQiu** [comment]
@calvinjia @uronce-cc PTAL, thanks!
Main Changes:
(1) cleanup S3A intermediate upload in leader master by periodically calling `UnderFileSystem.cleanup`.
(2) removed Fuse open() delay to PR https://github.com/Alluxio/alluxio/pull/7807
(3) flush() will be sent to worker and flush() will not happens before write() in worker side.

**Comment by LuQQiu** [comment]
@calvinjia The sync `flush()` trigger strange issue in `FileInStreamIntegrationTest` and I haven't find out the reason, so I roll back all the sync flush related changes in this PR. 
Make sync `flush()` another PR may be better to help find out the test failed reason. Thanks

**Comment by LuQQiu** [comment]
@calvinjia Added `S3ALowLevelOutputStreamTest` and fixed some bugs, PTAL, thanks

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java [hunk 19]
```diff
@@ -0,0 +1,422 @@
...
+import java.util.concurrent.atomic.AtomicInteger;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * [Experimental] A stream for writing a file into S3 using streaming upload.
+ * The data will be persisted to a temporary directory on the local disk
+ * and uploaded when the buffered data reaches the partition size. The data
+ * transfer is done using an S3 low-level multipart upload API.
+ */
```
Comment (calvinjia) [inline underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java]: Could you summarize the relationship between the methods and variables and the invariants? In particular, how the `init` methods manage the variables and the state of `mFile` and `mOffset` expected by each of the methods?

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java [hunk 20]
```diff
@@ -181,10 +195,9 @@ public static S3AUnderFileSystem createInstance(AlluxioURI uri,
     }
     clientConf.setMaxConnections(numThreads);
 
-    // Set client request timeout for all requests since multipart copy is used, and copy parts can
-    // only be set with the client configuration.
-    clientConf
-        .setRequestTimeout((int) Configuration.getMs(PropertyKey.UNDERFS_S3A_REQUEST_TIMEOUT));
+    boolean streamingUploadEnabled =
+        conf.isSet(PropertyKey.UNDERFS_S3A_STREAMING_UPLOAD_ENABLED)
+        && conf.getBoolean(PropertyKey.UNDERFS_S3A_STREAMING_UPLOAD_ENABLED);
```
Comment (calvinjia) [inline underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java]: Do we need to check `isSet` since there is a default value?

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java [hunk 21]
```diff
@@ -208,12 +221,13 @@ public static S3AUnderFileSystem createInstance(AlluxioURI uri,
         .fixedThreadPoolExecutorServiceFactory("alluxio-s3-transfer-manager-worker",
             numTransferThreads).create();
 
-    TransferManager transferManager = new TransferManager(amazonS3Client, service);
+    TransferManager transferManager = TransferManagerBuilder.standard()
+        .withS3Client(amazonS3Client).withExecutorFactory(() -> service)
+        .withMultipartCopyThreshold(MULTIPART_COPY_THRESHOLD)
+        .build();
```
Comment (calvinjia) [inline underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java]: Is this the recommended way of constructing the transfer manager?

Comment (LuQQiu) [reply underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java]: The old way has been deprecated so I change to use this way which the official doc recommends.

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java (hunk 20 above)
Comment (LuQQiu) [reply underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java]: I am a little confused with the configuration in `S3AUnderFileSystem`. Some configuration comes from the `UnderFileSystemConfiguration` and some come from the Global `Configuration`.

For here, `isSet` is not needed, I will remove it, thanks

File: core/common/src/main/java/alluxio/heartbeat/HeartbeatContext.java [hunk 22]
```diff
@@ -39,6 +39,8 @@
   public static final String MASTER_LOST_MASTER_DETECTION = "Master Lost Master Detection";
   public static final String MASTER_LOST_WORKER_DETECTION = "Master Lost Worker Detection";
   public static final String MASTER_METRICS_SYNC = "Master Metrics Sync";
+  public static final String MASTER_S3A_INTERMEDIATE_UPLOAD_CLEAN
```
[the remark this answers is not in the document]
Comment (LuQQiu) [reply core/common/src/main/java/alluxio/heartbeat/HeartbeatContext.java]: Renamed to 
```
MASTER_UFS_CLEANUP = "Master Ufs Cleanup";
```
Thanks

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java [hunk 23]
```diff
@@ -0,0 +1,422 @@
...
+  private boolean mClosed = false;
+
+  /**
+   * The local file that will be uploaded when reaches the partition size
+   * or when flush() is called and this file is bigger than 5MB.
+   */
+  private File mFile;
+
+  /** Store the future of tags. */
+  private List<ListenableFuture<PartETag>> mFutureTags = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (LuQQiu) [reply underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java]: Renamed, thanks!

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java [hunk 24]
```diff
@@ -0,0 +1,422 @@
...
+    // accept intermediate part smaller than 5 MB.
+    mPartitionSize = Math.max(UPLOAD_THRESHOLD,
+        Configuration.getBytes(PropertyKey.UNDERFS_S3A_STREAMING_UPLOAD_PARTITION_SIZE));
+    mPartNumber = new AtomicInteger(1);
+  }
+
+  @Override
+  public void write(int b) throws IOException {
+    mSingleCharWrite[0] = (byte) b;
+    write(mSingleCharWrite, 0, 1);
```
[the remark this answers is not in the document]
Comment (LuQQiu) [reply underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java]: Changed, thanks

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java [hunk 25]
```diff
@@ -0,0 +1,422 @@
...
+
+  @Override
+  public void write(byte[] b) throws IOException {
+    write(b, 0, b.length);
+  }
+
+  @Override
+  public void write(byte[] b, int off, int len) throws IOException {
+    if (mUploadId == null) {
+      initMultiPartUpload();
```
[the remark this answers is not in the document]
Comment (LuQQiu) [reply underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java]: Changed to
```
  @Override
  public void write(byte[] b, int off, int len) throws IOException {
    if (b == null || len == 0) {
      return;
    }
    validateWriteArgs(b, off, len);
    if (mUploadId == null) {
      initMultiPartUpload();
    }
    if (mFile == null) {
      initNewFile();
    }
```
do len and null check first, and then validate write arguments, see if any arguments out of range.
Finally init multipart upload or file.
Thanks

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java [hunk 26]
```diff
@@ -0,0 +1,422 @@
...
+  private List<ListenableFuture<PartETag>> mFutureTags = new ArrayList<>();
+
+  /** The MD5 hash of the file. */
+  private MessageDigest mHash;
+
+  /** The output stream to a local file where the file will be buffered. */
+  private OutputStream mLocalOutputStream;
+
+  /** When the offset reaches the partition size, we upload the temp file. */
+  private long mOffset;
```
[the remark this answers is not in the document]
Comment (LuQQiu) [reply underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java]: Agree, changed to `mPartitionOffset` to improve the correctness.

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java (hunk 19 above)
Comment (LuQQiu) [reply underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java]: Provide the following detailed relationship and explanation to streaming upload.
```
/**
 * [Experimental] A stream for writing a file into S3 using streaming upload.
 * The data transfer is done using S3 low-level multipart upload.
 *
 * The multipart upload is initialized in the first write() and an upload id is given
 * by AWS S3 to distinguish different multipart uploads.
 *
 * We upload data in partitions. When write(), the data will be persisted to
 * a temporary file {@link #mFile} on the local disk. When the data {@link #mPartitionOffset}
 * in this temporary file reaches the {@link #mPartitionSize}, the file will be submitted
 * to the upload executor {@link #mExecutor} and we do not wait for uploads to finish.
 * A new temp file will be created for the future write and the {@link #mPartitionOffset}
 * will be reset to zero. The process goes until all the Fdata has been written to temp files.
 *
 * In flush(), we upload the buffered data if they are bigger than 5MB
 * and wait for all uploads to finish. The temp files will be deleted after uploading successfully.
 *
 * In close(), we upload the last part of data (if exists), wait for all uploads to finish,
 * and complete the multipart upload.
 *
 * close() will not be retried, but all the multipart upload
 * related operations(init, upload, complete, and abort) will be retried.
 *
 * If an error occurs and we have no way to recover, we abort the multipart uploads.
 * Some multipart uploads may not be completed/aborted in normal ways and need periodical cleanup
 * by enabling the {@link PropertyKey#UNDERFS_CLEANUP_ENABLED}.
 * When a leader master starts or a cleanup interval is reached, all the multipart uploads
 * older than {@link PropertyKey#UNDERFS_S3A_INTERMEDIATE_UPLOAD_CLEAN_AGE} will be cleaned.
 */
```
Any recommendation for the above docs here?

**Comment by LuQQiu** [comment]
@calvinjia Thanks for the suggestions. I have addressed the issues. PTAL, thanks!

**Comment by LuQQiu** [comment]
The Streaming upload docs will be in another PR.

**Comment by calvinjia** [comment]
@LuQQiu Do you know why the proto files are changed?

**Comment by LuQQiu** [comment]
@calvinjia 
The proto files changed to include lineage deprecated information. Maybe because after we deprecate lineage, we do not regenerate the proto files. 

I first include sync flush in this PR and regenerate the proto files.
After that, I moved the sync flush to another PR, deleted the `WriteRequest.flush` in `protocol.proto` file and regenerate the proto files.

I think it is fine to include the proto file changes here since they only include lineage deprecate information.

**Comment by LuQQiu** [comment]
I removed the proto files and let sync flush PR do the lineage deprecated changes.
